### PR TITLE
Issue #1055: CatalogImportApiTest::getNextMessageFromQueue is an endl…

### LINKED ISSUE #1055: CatalogImportApiTest::getNextMessageFromQueue is an endless loop
If we don't add a message to the queue this is an endless loop.

Scrutinizer kills this after 180seconds, I think the test should kill it automatically after 5 seconds or something. No integration test runs longer.

And the second problem is: If one doesn't know what is happening currently they have no clue were to look for the problem.

### PR DESCRIPTION
**Pull request by Schrank**
…ess loop

If the tests doesn't stop, one doesn't know what the problem is, now we
have a helpful error message

Fixes #1055 

To test this comment out the body of \LizardsAndPumpkins\Messaging\Queue\InMemoryQueue::add

### CODE REVIEW

File: tests/Integration/Suites/CatalogImportApiTest.php [hunk 1]
```diff
@@ -26,11 +33,37 @@ public function receive(Message $message)
...
         return $receiver->message;
     }
 
+    protected function setUp()
+    {
+        parent::setUp();
+        $this->setupTimeOutToNoticeNoMessageAdded();
+    }
+
+    protected function setupTimeOutToNoticeNoMessageAdded()
```
Comment (Vinai) [inline tests/Integration/Suites/CatalogImportApiTest.php]: I think this should be `private`

File: tests/Integration/Util/Messaging/Queue/InMemoryQueue.php [hunk 2]
```diff
@@ -1,6 +1,7 @@
 <?php
 
 declare(strict_types=1);
+declare(ticks=1);
```
Comment (Vinai) [inline tests/Integration/Util/Messaging/Queue/InMemoryQueue.php]: Why is this needed here?

File: tests/Integration/Util/Messaging/Queue/InMemoryQueue.php [hunk 3]
```diff
@@ -53,6 +54,7 @@ public function consume(MessageReceiver $messageReceiver, int $numberOfMessagesT
                 $messageReceiver->receive($this->next());
                 $numberOfMessagesToConsumeBeforeReturn--;
             }
+            usleep(1);
```
Comment (Vinai) [inline tests/Integration/Util/Messaging/Queue/InMemoryQueue.php]: Why is this needed?

File: tests/Integration/Util/Messaging/Queue/InMemoryQueue.php (hunk 2 above)
Comment (Schrank) [inline tests/Integration/Util/Messaging/Queue/InMemoryQueue.php]: Because loop structures doesn't tick.

File: tests/Integration/Util/Messaging/Queue/InMemoryQueue.php (hunk 3 above)
Comment (Schrank) [inline tests/Integration/Util/Messaging/Queue/InMemoryQueue.php]: Loop structures and `if` doesn't tick, so we need a function call here.

**Comment by Schrank** [comment]
What about just adding a timeout to the InMemory implementation. It is only used for testing, so if it takes longer than 10 sec to add a message, we just break the loop and throw an exception. If a message is added, we reset the timeout?

**Comment by Vinai** [comment]
I don't like the idea to be honest.
This should not be part of the queues responsibility. It's not part of the interface contract.
If an endless loop occurs, then it's because of a bug that got introduced somewhere.
It will be pretty obvious to a developer when that happens since the tests won't complete.
I don't think there is anything wrong with expecting a developer to hit Ctrl-C or click "Stop" in the IDE to abort the tests.

File: tests/Integration/Util/Messaging/Queue/InMemoryQueue.php (hunk 2 above)
Comment (Vinai) [inline tests/Integration/Util/Messaging/Queue/InMemoryQueue.php]: When using this approach it worked with only declaring the `ticks=1` in the test in question.

File: tests/Integration/Util/Messaging/Queue/InMemoryQueue.php (hunk 3 above)
Comment (Vinai) [inline tests/Integration/Util/Messaging/Queue/InMemoryQueue.php]: Okay, that makes sense then.

**Comment by Schrank** [comment]
For me it wasn't obvious. I just copied the code, not knowing a queue with endless loop is involved.

Which is of course far away of ideal, nonetheless is copying test code without knowing implementation specifics normal :-)

So what about implementing an InMemoryQueueWithTimeout for integration test?

The alternative is to close this ticket, except you have another idea
